src/wp-admin/includes/class-rollback-auto-update.php

```diff
@@ -0,0 +1,440 @@
+<?php
+/**
+ * WordPress Plugin Administration API: WP_Rollback_Auto_Update class
+ *
+ * @package WordPress
+ * @subpackage Administration
+ * @since 6.3.0
+ */
+
+/**
+ * Core class for rolling back auto-update plugin failures.
+ */
+class WP_Rollback_Auto_Update {
+
+	/**
+	 * Stores handler parameters.
+	 *
+	 * @since 6.3.0
+	 *
+	 * @var array
+	 */
+	private $handler_args = array();
+
+	/**
+	 * Stores successfully updated plugins.
+	 *
+	 * @since 6.3.0
+	 *
+	 * @var array
+	 */
+	private static $processed = array();
+
+	/**
+	 * Stores fataling plugins.
+	 *
+	 * @since 6.3.0
+	 *
+	 * @var array
+	 */
+	private static $fatals = array();
+
+	/**
+	 * Stores `update_plugins` transient.
+	 *
+	 * @since 6.3.0
+	 *
+	 * @var stdClass
+	 */
+	private static $current;
+
+	/**
+	 * Stores boolean for no error from check_plugin_for_errors().
+	 *
+	 * @since 6.3.0
+	 *
+	 * @var bool
+	 */
+	private $update_is_safe = false;
+
+	/**
+	 * Stores error codes.
+	 *
+	 * @since 6.3.0
+	 *
+	 * @var int
+	 */
+	public $error_types = E_ERROR | E_PARSE | E_COMPILE_ERROR | E_USER_ERROR | E_RECOVERABLE_ERROR;
+
+	/**
+	 * Checks the validity of the updated plugin.
+	 *
+	 * @since 6.3.0
+	 *
+	 * @param array|WP_Error $result     Result from WP_Upgrader::install_package().
+	 * @param array          $hook_extra Extra arguments passed to hooked filters.
+	 * @return array|WP_Error
+	 */
+	public function auto_update_check( $result, $hook_extra ) {
+		if ( is_wp_error( $result ) || ! wp_doing_cron() || ! isset( $hook_extra['plugin'] ) ) {
+			return $result;
+		}
+
+		// Already processed.
+		if ( in_array( $hook_extra['plugin'], self::$processed, true ) ) {
+			return $result;
+		}
```

Comment:
Since a lookup is faster than `in_array()`, a slight performance improvement for bulk auto-updates could be:
```php
if ( isset( self::$processed[ $hook_extra['plugin'] ] ) ) {
```

Then instead of:
```php
self::$processed[] = $hook_extra['plugin'];
```
use:
```php
self::$processed[ $hook_extra['plugin'] ] = true;
```